### PR TITLE
[SPARK-20976] [SQL] Unify Error Messages for FAILFAST mode

### PR DESCRIPTION
**Pull request by gatorsmile**
### What changes were proposed in this pull request?
Before 2.2, we indicate the job was terminated because of `FAILFAST` mode.
```
Malformed line in FAILFAST mode: {"a":{, b:3}
```
If possible, we should keep it. This PR is to unify the error messages. 

### How was this patch tested?
Modified the existing messages.

### CODE REVIEW

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/json/JacksonParser.scala [hunk 1]
```diff
@@ -278,7 +278,7 @@ class JacksonParser(
       // We cannot parse this token based on the given data type. So, we throw a
       // RuntimeException and this exception will be caught by `parse` method.
       throw new RuntimeException(
-        s"Failed to parse a value for data type $dataType (current token: $token).")
+        s"Failed to parse a value for data type ${dataType.catalogString} (current token: $token).")
```
Comment (gatorsmile) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/json/JacksonParser.scala]: Since this message is external, we can make it more readable.

**Comment by gatorsmile** [comment]
cc @cloud-fan

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/FailureSafeParser.scala [hunk 2]
```diff
@@ -65,7 +66,8 @@ class FailureSafeParser[IN](
         case DropMalformedMode =>
           Iterator.empty
         case FailFastMode =>
-          throw e.cause
+          throw new SparkException("Malformed records are detected in record parsing. " +
+            s"Parse Mode: ${FailFastMode.name}. Reasons: ${e.getMessage}", e.cause)
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/FailureSafeParser.scala]: the `Reasons: ${e.getMessage}` is duplicated, the stack trace will include the cause exception.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/json/JsonSuite.scala [hunk 3]
```diff
@@ -1954,7 +1955,8 @@ class JsonSuite extends QueryTest with SharedSQLContext with TestJsonData {
           .json(path)
           .collect()
       }
-      assert(exceptionTwo.getMessage.contains("Failed to parse a value"))
+      assert(exceptionTwo.getMessage.contains(" Malformed records are detected in record " +
```
Comment (dongjoon-hyun) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/json/JsonSuite.scala]: nit: `" Malformed` -> `"Malformed` ?

**Comment by cloud-fan** [comment]
LGTM, merging to master!

**Pull request closed without merging**